File: NSIDC/AMSR_E_L2_Ocean_V06_200206190029_D.hdf.py
"""
Copyright (C) 2014 John Evans

This example code illustrates how to access and visualize an NSIDC AMSR-E
Ocean HDF-EOS2 Swath data file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python AMSR_E_L2_Ocean_V06_200206190029_D.hdf.py

The HDF file must be in your current working directory.
"""

import os

import matplotlib as mpl
import matplotlib.pyplot as plt
from mpl_toolkits.basemap import Basemap
import numpy as np

USE_NETCDF4 = False

FILE_NAME = 'AMSR_E_L2_Ocean_V06_200206190029_D.hdf'

# Identify the data field.
DATAFIELD_NAME = 'High_res_cloud'

if USE_NETCDF4:
    from netCDF4 import Dataset
    nc = Dataset(FILE_NAME)
    data = nc.variables[DATAFIELD_NAME][:].astype(np.float64)
    latitude = nc.variables['Latitude'][:]
    longitude = nc.variables['Longitude'][:]
    scale_factor = nc.variables[DATAFIELD_NAME].Scale
else:
    from pyhdf.SD import SD, SDC
    hdf = SD(FILE_NAME, SDC.READ)

    # Read dataset.
    data2D = hdf.select(DATAFIELD_NAME)
    data = data2D[:,:].astype(np.float64)

    # Read geolocation dataset.
    lat = hdf.select('Latitude')
    latitude = lat[:,:]
    lon = hdf.select('Longitude')
    longitude = lon[:,:]

    # Retrieve attributes.
    attrs = data2D.attributes(full=1)
    sfa=attrs["Scale"]
    scale_factor = sfa[0]        

# There is a wrap-around effect to deal with, as some of the swath extends
# eastward over the international dateline.  Adjust the longitude to avoid
# the swath being smeared.
longitude[longitude < -170] += 360

# Apply the fill value and scaling equation.
data[data == -9990] = np.nan
data = data * scale_factor
data = np.ma.masked_array(data, np.isnan(data))

units = "mm"
long_name = DATAFIELD_NAME

# Draw a polar stereographic projection using the low resolution coastline
# database.
m = Basemap(projection='cyl', resolution='l',
            llcrnrlat=-90, urcrnrlat=90,
            llcrnrlon=-170, urcrnrlon=190)
m.drawcoastlines(linewidth=0.5)
m.drawparallels(np.arange(-90, 91, 45), labels=[1, 0, 0, 0])
m.drawmeridians(np.arange(-180,181,45), labels=[0, 0, 0, 1])
m.pcolormesh(longitude, latitude, data, latlon=True)

cb = m.colorbar()
cb.set_label(units)

basename = os.path.basename(FILE_NAME)
plt.title('{0}\n{1}'.format(basename, long_name))
fig = plt.gcf()
pngfile = "{0}.py.png".format(basename)
fig.savefig(pngfile)

File: NSIDC/AMSR_E_L2_Ocean_V06_200206190029_D_hdf.m
%
%  This example code illustrates how to access and visualize NSIDC
%  AMSR_E Ocean  L2 HDF-EOS2 Swath file in MATLAB.
%
%  If you have any questions, suggestions, comments on this example, 
% please use the HDF-EOS Forum (http://hdfeos.org/forums). 
% 
%  If you would like to see an  example of any other NASA HDF/HDF-EOS
% data product that is not listed in the HDF-EOS  Comprehensive Examples 
% page (http://hdfeos.org/zoo), feel free to contact us at 
% eoshelp@hdfgroup.org or post it at the HDF-EOS Forum 
% (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
%  $matlab -nosplash -nodesktop -r AMSR_E_L2_Ocean_V06_200206190029_D_hdf
%
% Tested under: MATLAB R2017a
% Last updated: 2018-9-24


FILE_NAME='AMSR_E_L2_Ocean_V06_200206190029_D.hdf';
SWATH_NAME='Swath1';

% Open the HDF-EOS2 file.
file_id = hdfsw('open', FILE_NAME, 'rdonly');

% Open swath.
swath_id = hdfsw('attach', file_id, SWATH_NAME);

% Read dataset.
DATAFIELD_NAME='High_res_cloud';

[data, fail] = hdfsw('readfield', swath_id, DATAFIELD_NAME, [], [], []);
[lon, status] = hdfsw('readfield', swath_id, 'Longitude', [], [], []);
[lat, status] = hdfsw('readfield', swath_id, 'Latitude', [], [], []);


% Detach swath object.
hdfsw('detach', swath_id);
hdfsw('close', file_id);

% Convert the data to double type for plot.
data=double(data);
lon=double(lon);
lat=double(lat);

% Read attributes from the data field.
SD_id = hdfsd('start',FILE_NAME, 'rdonly');
DATAFIELD_NAME='High_res_cloud';

sds_index = hdfsd('nametoindex', SD_id, DATAFIELD_NAME);

sds_id = hdfsd('select',SD_id, sds_index);

% Read units from the data field.
units_index = hdfsd('findattr', sds_id, 'Unit');
[units, status] = hdfsd('readattr',sds_id, units_index);

% Read scale_factor from the data field.
scale_index = hdfsd('findattr', sds_id, 'Scale');
[scale, status] = hdfsd('readattr',sds_id, scale_index);
scale = double(scale);

% Terminate access to the corresponding data set.
hdfsd('endaccess', sds_id);

% Close the file.
hdfsd('end', SD_id);

% Replace the fill value with NaN.
data(data==-9990) = NaN;

% Multiply scale.
data = data*scale;

f = figure('Name', FILE_NAME, 'Renderer', 'zbuffer', ...
           'Position', [0,0,800,600], 'visible', 'off');

axesm('MapProjection','eqdcylin', ...
      'Frame','on','Grid','on', ...
      'MeridianLabel','on','ParallelLabel','on','MLabelParallel','south')
coast = load('coast.mat');

surfm(lat, lon, data);
colormap('Jet');
h=colorbar();

plotm(coast.lat,coast.long,'k');
tightmap;

title({FILE_NAME;'High res cloud'}, 'Interpreter', 'None', ...
    'FontSize',12,'FontWeight','bold');

set (get(h, 'title'), 'string', units, 'Interpreter', 'None', ...
                   'FontSize',12,'FontWeight','bold');

saveas(f,[FILE_NAME '.m.png']);
exit;


